File: plugin-sdk/typescript/src/api/profiler.ts
/**
 * Performance Profiler API
 *
 * @description Performance monitoring and profiling utilities for plugins.
 * Provides detailed metrics, tracing, and performance analysis tools.
 */

/**
 * Performance sample with timing data
 */
export interface PerformanceSample {
  /** Sample name */
  name: string;
  /** Start time (high-resolution) */
  startTime: number;
  /** End time (high-resolution) */
  endTime: number;
  /** Duration in milliseconds */
  duration: number;
  /** Memory usage at sample time */
  memory?: MemoryUsage;
  /** Additional metadata */
  metadata?: Record<string, unknown>;
}

/**
 * Memory usage information
 */
export interface MemoryUsage {
  /** Used heap size in bytes */
  usedHeapSize: number;
  /** Total heap size in bytes */
  totalHeapSize: number;
  /** Heap size limit in bytes */
  heapSizeLimit: number;
  /** External memory usage in bytes */
  external?: number;
}

/**
 * Performance bucket for aggregated metrics
 */
export interface PerformanceBucket {
  /** Bucket name (e.g., 'tool:search', 'hook:onAgentStep') */
  name: string;
  /** Number of samples */
  count: number;
  /** Total duration */
  totalDuration: number;
  /** Average duration */
  avgDuration: number;
  /** Minimum duration */
  minDuration: number;
  /** Maximum duration */
  maxDuration: number;
  /** 50th percentile (median) */
  p50: number;
  /** 95th percentile */
  p95: number;
  /** 99th percentile */
  p99: number;
}

/**
 * Performance report
 */
export interface PerformanceReport {
  /** Plugin ID */
  pluginId: string;
  /** Report generation time */
  generatedAt: Date;
  /** Report duration (ms) */
  duration: number;
  /** Total samples collected */
  totalSamples: number;
  /** Aggregated buckets */
  buckets: PerformanceBucket[];
  /** Memory snapshots */
  memorySnapshots: Array<{ timestamp: number; usage: MemoryUsage }>;
  /** Slow operations detected */
  slowOperations: SlowOperationEntry[];
}

/**
 * Slow operation entry
 */
export interface SlowOperationEntry {
  /** Operation name */
  name: string;
  /** Duration in milliseconds */
  duration: number;
  /** Threshold that was exceeded */
  threshold: number;
  /** Timestamp */
  timestamp: number;
  /** Stack trace if available */
  stack?: string;
}

/**
 * Profiler configuration
 */
export interface ProfilerConfig {
  /** Enable automatic profiling */
  enabled: boolean;
  /** Sample rate (0-1, default 1.0) */
  sampleRate: number;
  /** Maximum samples to keep */
  maxSamples: number;
  /** Slow operation threshold (ms) */
  slowThreshold: number;
  /** Enable memory tracking */
  trackMemory: boolean;
  /** Memory snapshot interval (ms) */
  memorySnapshotInterval: number;
}

/**
 * Performance Profiler API
 *
 * @remarks
 * Provides performance monitoring, tracing, and analysis tools
 * for understanding and optimizing plugin performance.
 *
 * @example
 * ```typescript
 * // Start profiling
 * context.profiler.start();
 *
 * // Profile a function
 * const result = await context.profiler.profile('fetchData', async () => {
 *   return await fetchData();
 * });
 *
 * // Get metrics
 * const report = context.profiler.generateReport();
 * console.log('Average fetch time:', report.buckets.find(b => b.name === 'fetchData')?.avgDuration);
 *
 * // Monitor slow operations
 * context.profiler.onSlowOperation((op) => {
 *   console.warn(`Slow: ${op.name} took ${op.duration}ms`);
 * });
 * ```
 */
export interface PluginProfilerAPI {
  /**
   * Start profiling session
   */
  start(): void;

  /**
   * Stop profiling session
   */
  stop(): void;

  /**
   * Check if profiler is active
   */
  isActive(): boolean;

  /**
   * Configure profiler
   *
   * @param config - Partial configuration to update
   */
  configure(config: Partial<ProfilerConfig>): void;

  /**
   * Get current configuration
   */
  getConfig(): ProfilerConfig;

  /**
   * Profile a function execution
   *
   * @param name - Profile name
   * @param fn - Function to profile
   * @param metadata - Optional metadata
   * @returns Function result
   */
  profile<T>(
    name: string,
    fn: () => T | Promise<T>,
    metadata?: Record<string, unknown>
  ): Promise<T>;

  /**
   * Start a manual timing
   *
   * @param name - Timer name
   * @returns Function to stop timing
   */
  startTiming(name: string): () => PerformanceSample;

  /**
   * Record a performance sample manually
   *
   * @param sample - Sample data
   */
  recordSample(sample: PerformanceSample): void;

  /**
   * Get all recorded samples
   *
   * @param name - Optional filter by name
   * @param limit - Maximum samples to return
   */
  getSamples(name?: string, limit?: number): PerformanceSample[];

  /**
   * Clear all recorded samples
   */
  clearSamples(): void;

  /**
   * Get aggregated metrics for a name
   *
   * @param name - Metric name
   */
  getMetrics(name: string): PerformanceBucket | null;

  /**
   * Get all aggregated metrics
   */
  getAllMetrics(): PerformanceBucket[];

  /**
   * Generate a performance report
   */
  generateReport(): PerformanceReport;

  /**
   * Export report as JSON
   */
  exportReport(): string;

  /**
   * Get current memory usage
   */
  getMemoryUsage(): MemoryUsage | null;

  /**
   * Take a memory snapshot
   */
  takeMemorySnapshot(): void;

  /**
   * Get memory snapshots
   */
  getMemorySnapshots(): Array<{ timestamp: number; usage: MemoryUsage }>;

  /**
   * Monitor for slow operations
   *
   * @param handler - Handler called when slow operation detected
   * @returns Unsubscribe function
   */
  onSlowOperation(handler: (entry: SlowOperationEntry) => void): () => void;

  /**
   * Set slow operation threshold
   *
   * @param threshold - Threshold in milliseconds
   */
  setSlowThreshold(threshold: number): void;

  /**
   * Get slow operations
   */
  getSlowOperations(): SlowOperationEntry[];

  /**
   * Mark a performance timeline event
   *
   * @param name - Event name
   * @param detail - Optional detail data
   */
  mark(name: string, detail?: unknown): void;

  /**
   * Measure between two marks
   *
   * @param name - Measurement name
   * @param startMark - Start mark name
   * @param endMark - End mark name
   */
  measureBetween(name: string, startMark: string, endMark: string): number | null;
}

File: plugin-sdk/python/cognia/profiler.py
"""
Performance Profiler API for Cognia Plugin SDK

Provides performance monitoring and profiling utilities for plugins.
Includes detailed metrics, tracing, and performance analysis tools.
"""

from abc import ABC, abstractmethod
from typing import Any, Callable, Dict, List, Optional

from .types import (
    PerformanceSample,
    MemoryUsage,
    PerformanceBucket,
    PerformanceReport,
    SlowOperationEntry,
    ProfilerConfig,
)


class ProfilerAPI(ABC):
    """
    Performance Profiler API for plugins.
    
    Provides performance monitoring, tracing, and analysis tools
    for understanding and optimizing plugin performance.
    
    Example:
        # Start profiling
        context.profiler.start()
        
        # Profile a function
        result = await context.profiler.profile('fetch_data', fetch_data)
        
        # Get metrics
        report = context.profiler.generate_report()
        bucket = next((b for b in report.buckets if b.name == 'fetch_data'), None)
        if bucket:
            print('Average fetch time:', bucket.avg_duration)
        
        # Monitor slow operations
        context.profiler.on_slow_operation(lambda op: 
            print(f'Slow: {op.name} took {op.duration}ms')
        )
    """
    
    @abstractmethod
    def start(self) -> None:
        """Start profiling session"""
        pass
    
    @abstractmethod
    def stop(self) -> None:
        """Stop profiling session"""
        pass
    
    @abstractmethod
    def is_active(self) -> bool:
        """Check if profiler is active"""
        pass
    
    @abstractmethod
    def configure(self, config: ProfilerConfig) -> None:
        """
        Configure profiler.
        
        Args:
            config: Configuration to update
        """
        pass
    
    @abstractmethod
    def get_config(self) -> ProfilerConfig:
        """Get current configuration"""
        pass
    
    @abstractmethod
    async def profile(self, name: str, fn: Callable, metadata: Optional[Dict[str, Any]] = None) -> Any:
        """
        Profile a function execution.
        
        Args:
            name: Profile name
            fn: Function to profile
            metadata: Optional metadata
            
        Returns:
            Function result
        """
        pass
    
    @abstractmethod
    def start_timing(self, name: str) -> Callable[[], PerformanceSample]:
        """
        Start a manual timing.
        
        Args:
            name: Timer name
            
        Returns:
            Function to stop timing and return sample
        """
        pass
    
    @abstractmethod
    def record_sample(self, sample: PerformanceSample) -> None:
        """
        Record a performance sample manually.
        
        Args:
            sample: Sample data
        """
        pass
    
    @abstractmethod
    def get_samples(self, name: Optional[str] = None, limit: Optional[int] = None) -> List[PerformanceSample]:
        """
        Get all recorded samples.
        
        Args:
            name: Optional filter by name
            limit: Maximum samples to return
        """
        pass
    
    @abstractmethod
    def clear_samples(self) -> None:
        """Clear all recorded samples"""
        pass
    
    @abstractmethod
    def get_metrics(self, name: str) -> Optional[PerformanceBucket]:
        """
        Get aggregated metrics for a name.
        
        Args:
            name: Metric name
        """
        pass
    
    @abstractmethod
    def get_all_metrics(self) -> List[PerformanceBucket]:
        """Get all aggregated metrics"""
        pass
    
    @abstractmethod
    def generate_report(self) -> PerformanceReport:
        """Generate a performance report"""
        pass
    
    @abstractmethod
    def export_report(self) -> str:
        """Export report as JSON"""
        pass
    
    @abstractmethod
    def get_memory_usage(self) -> Optional[MemoryUsage]:
        """Get current memory usage"""
        pass
    
    @abstractmethod
    def take_memory_snapshot(self) -> None:
        """Take a memory snapshot"""
        pass
    
    @abstractmethod
    def get_memory_snapshots(self) -> List[Dict[str, Any]]:
        """Get memory snapshots"""
        pass
    
    @abstractmethod
    def on_slow_operation(self, handler: Callable[[SlowOperationEntry], None]) -> Callable[[], None]:
        """
        Monitor for slow operations.
        
        Args:
            handler: Handler called when slow operation detected
            
        Returns:
            Unsubscribe function
        """
        pass
    
    @abstractmethod
    def set_slow_threshold(self, threshold: float) -> None:
        """
        Set slow operation threshold.
        
        Args:
            threshold: Threshold in milliseconds
        """
        pass
    
    @abstractmethod
    def get_slow_operations(self) -> List[SlowOperationEntry]:
        """Get slow operations"""
        pass
    
    @abstractmethod
    def mark(self, name: str, detail: Optional[Any] = None) -> None:
        """
        Mark a performance timeline event.
        
        Args:
            name: Event name
            detail: Optional detail data
        """
        pass
    
    @abstractmethod
    def measure_between(self, name: str, start_mark: str, end_mark: str) -> Optional[float]:
        """
        Measure between two marks.
        
        Args:
            name: Measurement name
            start_mark: Start mark name
            end_mark: End mark name
            
        Returns:
            Duration in milliseconds or None
        """
        pass


__all__ = [
    "ProfilerAPI",
]
